bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,348 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.dataflow;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Admin;
+import org.apache.hadoop.hbase.client.BufferedMutator;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Mutation;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.RegionLocator;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.RetriesExhaustedWithDetailsException;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.util.Pair;
+
+import com.google.bigtable.v1.BigtableServiceGrpc.BigtableService;
+import com.google.cloud.bigtable.hbase1_0.BigtableConnection;
+import com.google.cloud.dataflow.sdk.Pipeline;
+import com.google.cloud.dataflow.sdk.coders.Coder;
+import com.google.cloud.dataflow.sdk.coders.CoderRegistry;
+import com.google.cloud.dataflow.sdk.io.BoundedSource;
+import com.google.cloud.dataflow.sdk.options.PipelineOptions;
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.PDone;
+
+/**
+ * CloudBigtableIO is the gateway for Dataflow / Cloud Bigtable setup, {@link Sink}s and @{link
+ * {@link Source}s. NOTE: Sinks are ready, and Sources are coming soon.
+ */
+
+public class CloudBigtableIO {
+  /**
+   * A {@link BoundedSource} for a Bigtable {@link Table}, which is potentially filtered by a
+   * {@link Scan}.
+   */
+  public static class Source extends BoundedSource<Result> {
+    private static final long serialVersionUID = -5580115943635114126L;
+
+    private final CloudBigtableScanConfiguration configuration;
+
+    /**
+     * A {@link BoundedSource} for a Bigtable {@link Table} with a start/stop key range, along with
+     * a potential filter via a {@link Scan}.
+     */
+    private static class SourceWithKeys extends BoundedSource<Result> {
+      private static final long serialVersionUID = 2561066007121040429L;
+
+      private final CloudBigtableScanConfiguration configuration;
+      private final byte[] startKey;
```

Comment:
You use both startKey or startRow. I think we can use everywhere startRow, which is what the scanner uses. Or is Key a dataflow concept? would be nice to stick with one.